helix-term/src/commands/engine/scheme.rs

```diff
@@ -0,0 +1,2003 @@
+            if event == PromptEvent::Validate {
+                if let Err(e) = ENGINE.with(|x| {
+                    let mut args = args[1..]
+                        .iter()
+                        .map(|x| x.clone().into_steelval().unwrap())
+                        .collect::<Vec<_>>();
+
+                    let mut guard = x.borrow_mut();
+
+                    let res = {
+                        let mut cx = Context {
+                            register: None,
+                            count: std::num::NonZeroUsize::new(1),
+                            editor: cx.editor,
+                            callback: Vec::new(),
+                            on_next_key_callback: None,
+                            jobs: cx.jobs,
+                        };
+
+                        guard
+                            .with_mut_reference(&mut cx)
+                            .consume(|engine, mut arguments| {
+                                arguments.append(&mut args);
+
+                                engine.call_function_by_name_with_args(&parts[0], arguments)
+                            })
+                    };
+
+                    res
+                }) {
+                    compositor_present_error(cx, e);
+                };
+            }
+
+            // Global exists
+            true
+        } else {
+            // Global does not exist
+            false
+        }
+    }
+
+    fn get_doc_for_identifier(&self, ident: &str) -> Option<String> {
+        ExportedIdentifiers::engine_get_doc(ident)
+            .map(|v| v.into())
+            .or_else(|| {
+                if Self::is_exported(self, ident) {
+                    Some("Run this plugin command!".into())
+                } else {
+                    None
+                }
+            })
+    }
+
+    fn available_commands<'a>(&self) -> Vec<Cow<'a, str>> {
+        EXPORTED_IDENTIFIERS
+            .identifiers
+            .read()
+            .unwrap()
+            .iter()
+            .map(|x| x.clone().into())
+            .collect::<Vec<_>>()
+    }
+
+    fn load_theme(&self, name: &str) -> Option<helix_view::Theme> {
+        ThemeContainer::get(name)
+    }
+
+    fn themes(&self) -> Option<Vec<String>> {
+        let names = ThemeContainer::names();
+
+        if !names.is_empty() {
+            Some(names)
+        } else {
+            None
+        }
+    }
+
+    fn load_language_configuration(&self) -> Option<Result<Configuration, toml::de::Error>> {
+        LanguageConfigurationContainer::get_language_configuration()
+    }
+}
+
+impl SteelScriptingEngine {
+    fn is_exported(&self, ident: &str) -> bool {
+        EXPORTED_IDENTIFIERS
+            .identifiers
+            .read()
+            .unwrap()
+            .contains(ident)
+    }
+
+    // Attempt to fetch the keymap for the extension
+    fn get_keymap_for_extension<'a>(cx: &'a mut Context) -> Option<SteelVal> {
```

Comment:
this is too specical cased IMO.

While nice for a proof of concept the editor should have a generic concept of per buffer keymaps and then plugins should just interface with that

Reply:
I would agree, it requires a little hacking from the steel side to make it work as you can tell. Do you think it is worth investing in that machinery to expose that from the editor side now?